### PR TITLE
[SPARK-30070][SQL] Support ANSI datetimes predicate - overlaps

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?
This pull request introduces ANSI `OVERLAPS` predicate to SparkSQL. It yields true when two time periods (defined by their endpoints) overlap, false when they do not overlap.
Syntaxes as bellow.
```sql
(start1, end1) OVERLAPS (start2, end2)
(start1, length1) OVERLAPS (start2, length2)
```

```scala
 /**
 * The operator `OVERLAPS` determines whether or not two chronological periods overlap in time. A
 * chronological period is specified by a pair of datetimes (starting and ending) or as a
 * starting datetime and an interval(resolved by analyzer to datetimes pairs early).
 *
 * If the length of the period is greater than 0, then the period consists of all points of time
 * greater than or equal to the lower endpoint, and less than the upper endpoint,
 * a.k.a [lower, upper).
 *
 * If the length of the period is equal to 0, then the period consists of a single point in time,
 * the lower endpoint, a.k.a [lower, lower].
 *
 * Two periods overlap if they have at least one point in common.
 *
 */
```

```sql
postgres=# select (cast(a as timestamp), cast(b as timestamp)) overlaps (cast(c as timestamp), cast(d as timestamp)) from (values
 ('2011-11-11', '2011-11-11', '2011-11-11', '2011-11-11'),
 ('2011-11-10', '2011-11-11', '2011-11-11', '2011-11-12'),
 ('2011-11-11', '2011-11-10', '2011-11-11', '2011-11-12'),
 ('2011-11-11', '2011-11-10', '2011-11-12', '2011-11-11'),
 ('2011-11-10', '2011-11-11', '2011-11-12', '2011-11-13'),
 ('2011-11-10', '2011-11-20', '2011-11-11', '2011-11-19'),
 ('2011-11-11', '2011-11-19', '2011-11-10', '2011-11-20')) t(a,b,c,d);
 overlaps
----------
 t
 f
 f
 f
 f
 t
 t
(7 rows)
```

### Why are the changes needed?

ANSI support FeatureID: F053, and better PostgreSQL feature parity too.

### Does this PR introduce any user-facing change?

Yes, add a datetime predicate.

### How was this patch tested?

add ut

### CODE REVIEW

**Comment by yaooqinn** [comment]
retest this please

**Comment by maropu** [comment]
The behaivour of the other systems (e.g., oracle and mysql) is the same with the pg one above? Can you check?

**Comment by yaooqinn** [comment]
Hi @maropu, the `overlaps`'s behavior  in the SQL standard is very clear, and PostgreSQL follows it exactly.  So here we do exactly the same.
In Oracle, it is also supported but undocumented. 
Not supported by MySQL.

**Comment by maropu** [comment]
I'm not faimilar with this syntax though, to regard the range as `[lower, upper)` is the same among them? Which the description below does come from, the pg doc or the SQL definition?
```
 /**
 * The operator `OVERLAPS` determines whether or not two chronological periods overlap in time. A
 * chronological period is specified by a pair of datetimes (starting and ending) or as a
 * starting datetime and an interval(resolved by analyzer to datetimes pairs early).
 *
 * If the length of the period is greater than 0, then the period consists of all points of time
 * greater than or equal to the lower endpoint, and less than the upper endpoint,
 * a.k.a [lower, upper).
 *
 * If the length of the period is equal to 0, then the period consists of a single point in time,
 * the lower endpoint, a.k.a [lower, lower].
 *
 * Two periods overlap if they have at least one point in common.
 *
 */
```

**Comment by yaooqinn** [comment]
The comments above are based on the SQL Standard.
Here is the pg's description
> This expression yields true when two time periods (defined by their endpoints) overlap, false when they do not overlap. The endpoints can be specified as pairs of dates, times, or time stamps; or as a date, time, or time stamp followed by an interval. When a pair of values is provided, either the start or the end can be written first; OVERLAPS automatically takes the earlier value of the pair as the start. Each time period is considered to represent the **half-open interval start <= time < end**, unless start and end are equal in which case it represents that single time instant. This means for instance that two time periods with only an endpoint in common do not overlap.
from link:
https://www.postgresql.org/docs/current/functions-datetime.html

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala [hunk 1]
```diff
@@ -2165,3 +2165,75 @@ case class SubtractDates(left: Expression, right: Expression)
...
+          If(EqualTo(rightStart, rightEnd), /** right period is 0, rl ∈ [ll, lu) */
+            And(GreaterThanOrEqual(rightStart, Least(left)), LessThan(rightStart, Greatest(left))),
+            {
+              val timePoints = Seq(leftStart, leftEnd, rightStart, rightEnd)
+              val leftPeriod = If(GreaterThan(leftStart, leftEnd),
+                Subtract(leftStart, leftEnd), Subtract(leftEnd, leftStart))
+              val rightPeriod = If(GreaterThan(rightStart, rightEnd),
+                Subtract(rightStart, rightEnd), Subtract(rightEnd, rightStart))
+              LessThan(Subtract(Greatest(timePoints), Least(timePoints)),
+                Add(leftPeriod, rightPeriod))
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala]: Can you do this replacement directly in `ResolveDateTimeOverlaps`?

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala]: Yes but can you tell me why？ Methinks it is better we have a override def sql here.

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala]: Ah, I see. Better to keep the current one.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 2]
```diff
@@ -246,6 +248,29 @@ class Analyzer(
...
 
+  /**
+   * Replace interval with startEnd + interval and substitute child plan with DateTimeOverlaps.
+   */
+  object ResolveDateTimeOverlaps extends Rule[LogicalPlan] {
+    def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsUp {
+      case p: LogicalPlan => p.transformAllExpressions {
+        case u @ UnresolvedDateTimeOverlaps(leftStart, leftEnd, _, _)
+          if leftEnd.dataType.isInstanceOf[CalendarIntervalType] =>
+          u.copy(leftEnd = Add(leftStart, leftEnd))
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Why did you do nothing for right here? We cannot replace it with `DateTimeOverlaps` in this match?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 3]
```diff
@@ -246,6 +248,29 @@ class Analyzer(
...
   )
 
+  /**
+   * Replace interval with startEnd + interval and substitute child plan with DateTimeOverlaps.
+   */
+  object ResolveDateTimeOverlaps extends Rule[LogicalPlan] {
+    def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsUp {
+      case p: LogicalPlan => p.transformAllExpressions {
+        case u @ UnresolvedDateTimeOverlaps(leftStart, leftEnd, _, _)
+          if leftEnd.dataType.isInstanceOf[CalendarIntervalType] =>
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: `.sameType(CalendarIntervalType)`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 2 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: SGTM. I’ll refine this

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 3 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: changed this but with `match-case`

**Comment by yaooqinn** [comment]
retest this please

**Comment by yaooqinn** [comment]
cc @cloud-fan @gatorsmile, thanks for reviewing this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 4]
```diff
@@ -246,6 +248,32 @@ class Analyzer(
...
 
+  /**
+   * Replace interval with startEnd + interval and substitute child plan with DateTimeOverlaps.
+   */
+  object ResolveDateTimeOverlaps extends Rule[LogicalPlan] {
+    def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsUp {
+      case p: LogicalPlan => p.transformAllExpressions {
+        case u @ UnresolvedDateTimeOverlaps(leftStart, leftEnd, rightStart, rightEnd) =>
+          withPosition(u) {
+            u.checkInputDataTypes() match {
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: We cannot use `ExpectsInputTypes` instead?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 5]
```diff
@@ -246,6 +248,32 @@ class Analyzer(
...
   )
 
+  /**
+   * Replace interval with startEnd + interval and substitute child plan with DateTimeOverlaps.
+   */
+  object ResolveDateTimeOverlaps extends Rule[LogicalPlan] {
+    def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsUp {
+      case p: LogicalPlan => p.transformAllExpressions {
+        case u @ UnresolvedDateTimeOverlaps(leftStart, leftEnd, rightStart, rightEnd) =>
+          withPosition(u) {
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: I think we don't need this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 4 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: no, we cannot. It‘s `RuntimeReplaceable`.

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: no, we cannot. It‘s `RuntimeReplaceable`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala (hunk 1 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala]: On second thought, this replacement looks difficult to me in terms of code maintainability. I tried to implement the overlap w/o `RuntimeReplaceable` like this: https://github.com/maropu/spark/commit/4e30922822fac497950bafd150929f02577ab5a4
The implementation and the generated code looks simpler. Probably, we need more comments about the approach from the other's reviewers. cc: @cloud-fan @viirya

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala]: looks nice

**Pull request closed without merging**